flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -814,6 +820,69 @@ public Executor getFutureExecutor() {
 				entry -> serializeAccumulator(entry.getKey(), entry.getValue())));
 	}
 
+	@Override
+	public BlockingPersistentResultPartitionMeta getBlockingPersistentResultPartitionMeta() {
+		Map<AbstractID, Map<AbstractID, ResultPartitionDescriptor>> resultPartitionDescriptors = new HashMap<>();
+
+		// keep record of all failed IntermediateDataSetID
+		Set<AbstractID> failedIntermediateDataSetIds = new HashSet<>();
+
+		for (ExecutionVertex executionVertex : getAllExecutionVertices()) {
+			for (IntermediateResultPartition intermediateResultPartition : executionVertex.getProducedPartitions().values()) {
+				if (intermediateResultPartition.getResultType() == ResultPartitionType.BLOCKING_PERSISTENT) {
+					try {
+						addLocation(resultPartitionDescriptors, intermediateResultPartition);
+					} catch (Throwable throwable) {
+						LOG.error("Failed to Failed to get location of ResultPartition: " + intermediateResultPartition.getPartitionId(), throwable);
+						failedIntermediateDataSetIds.add(
```

Comment:
When getting FailedIntermediateDataSets, we should also return the corresponding error messages so that the client side can be aware of the reason why they can not use the cache.